tests/test_tools.py

```diff
@@ -145,3 +152,345 @@ def test_unbound_tuple():
+
+
+class TestMemoizedInstances:
+    """
+    Tests for the `memoized_instances` decorator.
+    """
+
+    def test_memo(self):
+        """
+        Tests basic functionality of memoized instances.
+        """
+        @_memoized_instances
+        class Box:
+            def __init__(self, value: int):
+                self.value = value
+                self.init_calls = getattr(self, 'init_calls', 0) + 1
+
+        # Create instances with the same value
+        box1 = Box(10)
+        box2 = Box(10)
+        box3 = Box(20)
+
+        # Ensure they are the same instance
+        assert box1.value == 10
+        assert box1 is box2
+
+        # Ensure initialization only happened once
+        assert box1.init_calls == 1
+
+        # Ensure different values create different instances
+        assert box1 is not box3
+        assert box3.init_calls == 1
+
+    def test_memo_with_new(self):
+        """
+        Tests that `memoized_instances` works correctly with `__new__`.
+        """
+        @_memoized_instances
+        class BoxWithNew:
+            def __new__(cls, value: int):
+                instance = super().__new__(cls)
+                instance.value = value
+                return instance
+
+            def __init__(self, value: int):
+                self.value += value
+                self.init_calls = getattr(self, 'init_calls', 0) + 1
+
+        # Create instances with the same value
+        box1 = BoxWithNew(10)
+        box2 = BoxWithNew(10)
+        box3 = BoxWithNew(20)
+
+        # Ensure they are the same instance
+        assert box1.value == 20
+        assert box1 is box2
+
+        # Ensure initialization only happened once
+        assert box1.init_calls == 1
+
+        # Ensure different values create different instances
+        assert box1 is not box3
+        assert box3.init_calls == 1
+
+    def test_subclass_memo(self):
+        """
+        Tests that applying the decorator multiple times in an inheritance chain
+        does not change the behavior.
+        """
+        @_memoized_instances
+        # @_memoized_instances
+        class Box:
+            def __init__(self, value: int):
+                self.value = value
+                self.init_calls = getattr(self, 'init_calls', 0) + 1
+
+        @_memoized_instances
+        class SubBox(Box):
+            def __init__(self, value: int):
+                super().__init__(value)
+                self.sub_init_calls = getattr(self, 'sub_init_calls', 0) + 1
+
+        # Create instances with the same value
+        box = Box(10)
+        subbox1 = SubBox(10)
+        subbox2 = SubBox(10)
+        subbox3 = SubBox(20)
+
+        # Ensure the subclass instances are not the same as the base class
+        assert box is not subbox1
+        assert subbox1 is subbox2
+        assert subbox1 is not subbox3
+
+    def test_subclass_missing_decorator(self):
+        """
+        Tests that not applying the decorator to a child class raises an error.
```

Comment:
That sounds quite problematic to force subclass to be moized instances too. What if the subclass creates objects to big to be cached so we only want to allow the parent to memorize?